crate/utils/src/crypto/hybrid_encryption/ecies.rs

```diff
@@ -0,0 +1,204 @@
+    let S_vec = S.to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
+
+    let mut key = vec![0; AES_256_GCM_KEY_LENGTH];
+
+    let mut hasher = Hasher::new(MessageDigest::shake_128())?;
+    hasher.update(&S_vec)?;
+    hasher.finish_xof(&mut key)?;
+
+    Ok(key)
+}
+
+/// Encrypt `plaintext` data using `pubkey` public key following the ECIES
+/// principle.
+///
+/// Generate a random `r` and compute `R = rG` with `G` the curve generator.
+/// Using target pubic key `pubkey` we will call `Q`, compute `S = rQ`. `S` is
+/// the shared key used to symmetrically encrypt data using AES-256-GCM.
+///
+/// Return `R | ct | tag` with `|` the concatenation operator, `R` the ephemeral
+/// public key on the curve, `ct` the encrypted data and `tag` the
+/// authentication tag forged during encryption.
+///
+/// Notice we don't send the IV since it is derived by hashing the public key as
+/// well as the ephemeral public key.
+#[allow(non_snake_case)]
+pub fn ecies_encrypt(pubkey: &PKey<Public>, plaintext: &[u8]) -> Result<Vec<u8>, KmipUtilsError> {
+    let mut ctx = BigNumContext::new_secure()?;
+    let Q = pubkey.ec_key()?;
+    let curve = Q.group();
+
+    #[cfg(feature = "fips")]
+    if curve.curve_name() == Some(Nid::X9_62_PRIME192V1) {
+        kmip_utils_bail!("Curve P-192 not allowed in FIPS mode.")
+    }
+
+    // Generating random ephemeral private key `r` and associated public key
+    // `R`.
+    let r = EcKey::generate(curve)?;
+    let R = EcKey::from_public_key(curve, r.public_key())?;
+
+    // Compute secret key from recipient public key `S = rQ`.
+    let mut S = EcPoint::new(curve)?;
+    S.mul(curve, Q.public_key(), r.private_key(), &ctx)?;
+
+    let key = ecies_get_key(&S, curve)?;
+    let iv = ecies_get_iv(Q.public_key(), R.public_key(), curve)?;
+
+    // Allocating memory for AES-GCM to write the tag at.
+    let mut tag = vec![0; AES_256_GCM_MAC_LENGTH];
+
+    // Encryption using AES-256-GCM.
+    let ct: Vec<u8> = encrypt_aead(
+        Cipher::aes_256_gcm(),
+        &key,
+        Some(&iv),
+        &[],
+        plaintext,
+        tag.as_mut(),
+    )?;
+
+    let R_vec = R
+        .public_key()
+        .to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
+
+    Ok([R_vec, ct, tag].concat())
+}
+
+/// Decrypt `ciphertext` data using `privkey` private key following the ECIES
+/// principle.
+///
+/// `ciphertext` is a concatanation of `R | ct | tag` with `|` the concatenation
+/// operator, `R` the ephemeral public key on the curve, `ct` the encrypted data
+/// and `tag` the authentication tag forged during encryption.
+///
+/// The IV for decryption is computed by taking the hash of the recipient public
+/// key and the ephemeral public key.
+///
+/// Return the plaintext.
+#[allow(non_snake_case)]
+pub fn ecies_decrypt(
+    privkey: &PKey<Private>,
+    ciphertext: &[u8],
+) -> Result<Vec<u8>, KmipUtilsError> {
+    let mut ctx = BigNumContext::new_secure()?;
+    let d = privkey.ec_key()?;
```

Comment:
Why `d`? Maybe use shadowing:
```suggestion
    let privkey = privkey.ec_key()?;
```
Can this be done without copying?